postmortems/2023-03-20_azure_dev_outage.md

```diff
@@ -0,0 +1,54 @@
+
+Original deployment was intended as a proof of concept, it's a 'dev' resource group which had a planned migration to another environment. Since then the plan for productionising has been revised, and proper domain was given to the site to enable early testing with some groups within the client organisation, which give the appearance of a more mature solution. This means both monitoring and processes around management of the environment are less robust than for the primary deployment in our own setup.
+
+The original method for obtaining the secret was not well understood by us or the supplier, we originally walked one of their support agents through the process in the Azure portal on a video call, and were then emailed the result. That secret had lifetime of one year (likely a default) and was included in one of the emails, but not added to a calendar or otherwise recorded.
+
+There is no process in the client organisation for updating secrets. Neither Martin or JJ had ability to file the ticket under the options available, and had to go through helpdesk, and the ticket had to be filed as a "Non Standard Request". There were then multiple layers within the client organisation to navigate to get the supplier to look at this outside their 10 day window for work items not otherwise categorised. With chasing, it took three days for a response.
+
+## Wonderings
+
+There was no monitoring to detect the outage, it was only spotted when trying to access the service during working hours. Again due to having been created as a temporary deployment, and also restrictions from the client around our standard tooling. The web service exposes Prometheus metrics but we don't have a service principal would let us scrape, dashboard, and alert on them. We need to decide on an approach here as part of productionisation that would catch auth specific outages. Whitebox monitoring would not be sufficient.
+
+The component serving the 500 response (the easyauth 'sidecar' component) did not return a response with any useful details as to why it was failing, nor log anywhere that could be seen from within the azure portal or by sshing into an app service. This made discovering why authentication was failing difficult. At some point, we may have to switch to handling oauth within our web server for better observability and flexibility on handling.
+
+Our approach for infrequent non-automatable processes, such as awkward cert updates, is to have a documented playbook and a calendar invite reminder. Any deployment secrets required for Azure will need to use this framework.
+
+For something that prevented all use of our service within the client, it was unclear what level of urgency was actually appropriate. Once it was clear that we would be unable to resolve the issue without going through another supplier, we did not take follow up steps to verify how to resolve the problem using our own environment, which meant additional delay once the secret was eventually provided.
```

Comment:
Is this accurate? We did spin up our own environment, and we did verify that the expired secret was the cause (if only accidentally due to our own app registration secret also having expired), and we did verify that refreshing the secret fixed it.

Reply:
I guess the point is we didn't quite do that to the level of rigour that would have made the work following getting the new secret on Thursday well understodd - I wasn't aware it required a redeploy until seeing a reboot with after key vault change had no effect.